apps/apps.c

```diff
@@ -2851,3 +2970,159 @@ void make_uppercase(char *string)
+            || !BIO_meth_set_gets(methods_tempfile, tempfile_gets)
+            || !BIO_meth_set_ctrl(methods_tempfile, tempfile_ctrl)
+            || !BIO_meth_set_create(methods_tempfile, tempfile_new)
+            || !BIO_meth_set_destroy(methods_tempfile, tempfile_free))
+            return NULL;
+    }
+    return methods_tempfile;
+}
+
+static int tempfile_new(BIO *b)
+{
+    TEMPFILE_CTX *temp;
+
+    temp = app_malloc(sizeof(*temp), "temp file");
+    temp->filename = NULL;
+    temp->tmp_filename = NULL;
+    BIO_set_data(b, temp);
+    BIO_set_init(b, 1);
+
+    return 1;
+}
+
+static int tempfile_free(BIO *b)
+{
+    TEMPFILE_CTX *temp;
+
+    if (b == NULL)
+        return 0;
+
+    temp = BIO_get_data(b);
+    if (temp->filename != NULL && temp->tmp_filename != NULL) {
+#if !defined(_WIN32) && !defined(__VMS)
+        rename(temp->tmp_filename, temp->filename);
+#endif
+    }
+    if (temp->filename)
+        OPENSSL_free(temp->filename);
+    if (temp->tmp_filename)
+        OPENSSL_free(temp->tmp_filename);
+    OPENSSL_free(temp);
+    BIO_set_data(b, NULL);
+    BIO_set_init(b, 0);
+
+    return 1;
+}
+
+static int tempfile_read(BIO *b, char *out, int outl)
+{
+    int ret = 0;
+    BIO *next = BIO_next(b);
+
+    if (out == NULL || outl == 0)
+        return 0;
+    if (next == NULL)
+        return 0;
+
+    ret = BIO_read(next, out, outl);
+    return ret;
+}
+
+static int tempfile_write(BIO *b, const char *in, int inl)
+{
+    BIO *next = BIO_next(b);
+
+    if ((in == NULL) || (inl <= 0))
+        return 0;
+    if (next == NULL)
+        return 0;
+
+    return BIO_write(next, in, inl);
+}
+
+static long tempfile_ctrl(BIO *b, int cmd, long num, void *ptr)
+{
+    long ret = 1;
+    BIO *next = BIO_next(b);
+    TEMPFILE_CTX *temp;
+    char **p;
+
+    if (next == NULL)
+        return 0;
+    switch (cmd) {
+    case BIO_C_SET_FILENAME:
+        temp = BIO_get_data(b);
+        if (num == 1)
+            p = &temp->filename;
+        else
+            p = &temp->tmp_filename;
+        if (*p)
+            OPENSSL_free(*p);
+        *p = OPENSSL_strdup((char *)ptr);
+        if (p == NULL)
+            return 0;
+        break;
+    default:
+        ret = BIO_ctrl(next, cmd, num, ptr);
+        break;
+    }
+    return ret;
+}
+
+static int tempfile_gets(BIO *b, char *buf, int size)
+{
+    int i, ret = 0;
+    BIO *next = BIO_next(b);
+
+    if (next == NULL)
+        return 0;
+    for (i = 0; i < size - 1; ++i) {
+        ret = tempfile_read(b, &buf[i], 1);
+        if (ret <= 0)
+            break;
+        else if (buf[i] == '\n') {
+            ++i;
+            break;
+        }
+    }
+    if (i < size)
+        buf[i] = '\0';
+    return (ret < 0 && i == 0) ? ret : i;
+}
+
+static int tempfile_puts(BIO *b, const char *str)
+{
+    BIO *next = BIO_next(b);
+
+    if (next == NULL)
+        return 0;
+    return tempfile_write(next, str, strlen(str));
+}
```

Comment:
Frankly, we've already stuffed `apps.c` too much.  Please put the filter in a file of its own and add the file name [here](https://github.com/openssl/openssl/blob/master/apps/build.info#L11)

Reply:
(`bf_temp.c`, for example)

Reply:
Okay